From: Suren Baghdasaryan <surenb@google.com>
Subject: mm/mremap: write-lock VMA while remapping it to a new address range
Date: Mon, 27 Feb 2023 09:36:16 -0800

Write-lock VMA as locked before copying it and when copy_vma produces a
new VMA.

Link: https://lkml.kernel.org/r/20230227173632.3292573-18-surenb@google.com
Signed-off-by: Suren Baghdasaryan <surenb@google.com>
Reviewed-by: Laurent Dufour <laurent.dufour@fr.ibm.com>
Reviewed-by: Hyeonggon Yoo <42.hyeyoo@gmail.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---


--- a/mm/mmap.c~mm-mremap-write-lock-vma-while-remapping-it-to-a-new-address-range
+++ a/mm/mmap.c
@@ -3183,6 +3183,7 @@ struct vm_area_struct *copy_vma(struct v
 			get_file(new_vma->vm_file);
 		if (new_vma->vm_ops && new_vma->vm_ops->open)
 			new_vma->vm_ops->open(new_vma);
+		vma_start_write(new_vma);
 		if (vma_link(mm, new_vma))
 			goto out_vma_link;
 		*need_rmap_locks = false;
--- a/mm/mremap.c~mm-mremap-write-lock-vma-while-remapping-it-to-a-new-address-range
+++ a/mm/mremap.c
@@ -623,6 +623,7 @@ static unsigned long move_vma(struct vm_
 			return -ENOMEM;
 	}
 
+	vma_start_write(vma);
 	new_pgoff = vma->vm_pgoff + ((old_addr - vma->vm_start) >> PAGE_SHIFT);
 	new_vma = copy_vma(&vma, new_addr, new_len, new_pgoff,
 			   &need_rmap_locks);
_
